ccp.c

```diff
@@ -246,42 +250,48 @@ int ccp_read_msg(
         PRINT("unknown connection: %u\n", hdr.SocketId);
         return -5;
     }
+    msg_ptr = buf + ok;
 
     state = get_ccp_priv_state(conn);
     if (hdr.Type == INSTALL_EXPR) {
-        memset(&emsg, 0, sizeof(struct InstallExpressionMsg));
-        ok = read_install_expr_msg(&hdr, &emsg, buf + ok);
+        memset(&expr_msg_info, 0, sizeof(struct InstallExpressionMsgHdr));
+        ok = read_install_expr_msg_hdr(&hdr, &expr_msg_info, msg_ptr); 
         if (ok < 0) {
-            PRINT("could not read install expression msg: %d\n", ok);
+            PRINT("could not read install expression msg header: %d\n", ok);
             return -6;
         }
+        msg_ptr += ok;
 
         ACQUIRE_LOCK(&ccp_state_lock);
         memset(state->expressions, 0, MAX_EXPRESSIONS * sizeof(struct Expression));
         memset(state->fold_instructions, 0, MAX_INSTRUCTIONS * sizeof(struct Instruction64));
     
-        state->program_uid = emsg.program_uid;
-        state->num_expressions = emsg.num_expressions;
-        state->num_instructions = emsg.num_instructions;
+        state->program_uid = expr_msg_info.program_uid;
+        state->num_expressions = expr_msg_info.num_expressions;
+        state->num_instructions = expr_msg_info.num_instructions;
 
         // parse the expressions 
         for (i=0; i<state->num_expressions; i++) {
```

Comment:
Could this entire for loop instead just be:

```c
memcpy(state->expressions, msg_ptr, emsg.num_expressions * sizeof(struct ExpressionMsg))
```

?

Reply:
Yeah, I think it was this way for debugging purposes but this is more efficient.

Reply:
Actually - we can't do this right now. For some reason, portus sends down expression messages in u8s, whereas they are stored as 4 u32s (I don't know if there is a specific reason for this). @akshayknarayan ?

Reply:
That’s just how it’s dealt with in the code, it’s the same thing in memory

Reply:
Dealing with u8s in rust is equivalent to dealing with chars in C

Reply:
I'm not following. If portus sends down 4 u8s, and we say memcpy(state, buffer, sizeof (4 u32s)), it's going to copy stuff past the expression right?

Reply:
well, when I tried changing it to  a memcpy, the unit tests didn't pass.

Reply:
@deeptir18 not sure what you mean - https://github.com/ccp-project/portus/blob/master/src/serialize/install.rs#L13 is composed of u32s?

Reply:
I'm talking about an Expression Struct itself. If you look in https://github.com/ccp-project/libccp/blob/master/serialize.h#L139, there are 4 u8s in an expression. We can't memcpy this into a struct with 4 u32s directly (as a group) right?

Reply:
I looked more and I see what you are saying, let's discuss offline